### PR TITLE
Feat/backend setup

### PR DESCRIPTION
**Pull request by draskonovakovic**
Initial Backend Implementation

This pull request introduces the foundational backend for the project, fully implemented in TypeScript. It includes:

    User Management: CRUD functionality with password hashing using bcrypt for secure storage.
    Validation Middleware: Ensures user input integrity (e.g., name, surname, email, and password validation).
    Database Integration: PostgreSQL setup with a reusable query interface and repository pattern for modular and scalable data access.
    WebSocket Support: Real-time messaging via socket.io.
    Error Handling: Centralized error creation and global middleware for consistent responses.
    Clean Architecture: Organized structure with services, controllers, repositories, and models for extensibility. 

This pull request should set a solid foundation for future features and improvements.

@milan-at-2c @ivan-at-2c

### CODE REVIEW

File: src/repositories/baseRepository.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    async findById(id: number): Promise<T | null> {
+      const query = `SELECT * FROM ${table} WHERE id = $1`;
+      const result = await db.query(query, [id]);
+      return result.rows[0] || null;
+    },
+  
+    async update(id: number, data: Partial<T>): Promise<T | null> {
+      const fields = Object.keys(data).map((key, index) => `${key} = $${index + 1}`).join(', ');
+      const values = Object.values(data);
+      const query = `UPDATE ${table} SET ${fields} WHERE id = $${values.length + 1} RETURNING *`;
```
Comment (milan-at-2c) [inline src/repositories/baseRepository.ts:27]: values.length predstavlja broj elemenata u nizu, on ce relativno uvek biti isti. Sta ce biti kad ti je id korisnika 128? Svakako ovakva funkcionalnosti ne radi

File: src/repositories/baseRepository.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  
+      const result = await db.query(query, [...values, id]);
+      return result.rows[0] || null;
+    },
+  
+    async delete(id: number): Promise<boolean> {
+      const query = `DELETE FROM ${table} WHERE id = $1 RETURNING id`;
+      const result = await db.query(query, [id]);
+      return result.rowCount! > 0;
```
Comment (milan-at-2c) [inline src/repositories/baseRepository.ts]: srediti tipove koje vraca query, ne treba koristiti operator -> !

File: src/services/userService.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import { User } from '../models/user';
+import { createError } from '../utilis/createError';
+import bcrypt from 'bcrypt'
+import { config } from '../config/config';
+  
+export const userService = {
+    async createUser(user: Omit<User, 'id'>): Promise<User> {
+      const existingUser = await userRepository.findUserByEmail(user.email);
+      if (existingUser) {
+        throw createError('User with this email already exists', 409);
```
Comment (milan-at-2c) [inline src/services/userService.ts]: promeniti na 400, korisnikova je greska sto je pokusao sa vec postojecim mejlom

File: src/services/userService.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      }
+
+      const hashedPassword = await bcrypt.hash(user.password, config.bcrypt.saltRounds);
+
+      const newUser = { ...user, password: hashedPassword } as User;
```
Comment (milan-at-2c) [inline src/services/userService.ts]: Sto kastujes ovde?

File: package.json [hunk 5]
```diff
@@ -1,19 +1,28 @@
 {
   "dependencies": {
+    "bcrypt": "^5.1.1",
     "body-parser": "^1.20.3",
     "cors": "^2.8.5",
     "dotenv": "^16.4.5",
-    "express": "^4.21.1",
+    "express": "^4.18.2",
```
Comment (milan-at-2c) [inline package.json:7]: da li postoji razlog sto si snizio verziju?

**Review by milan-at-2c — CHANGES_REQUESTED** [review]

File: package.json (hunk 5 above)
Comment (draskonovakovic) [reply package.json:7]: Predpostavljam da je zbog nekog paketa koji sam instalirao, ne mogu da se setim tacno kojeg.

File: src/services/userService.ts (hunk 4 above)
Comment (draskonovakovic) [reply src/services/userService.ts]: Izbacio sam, vidim da nema potrebe. Nisam bio siguran kad sam kucao pa sam stavio za svaki slucaj.

**Review by milan-at-2c — APPROVED** [review]

**Pull request merged**